hilti/src/rt/tests/stream.cc

```diff
@@ -16,169 +16,471 @@ using hilti::rt::to_string;
-    i += 7;
-    CHECK(*i == '3');
-    i += 7;
-    CHECK(*i == '0');
-    i += 1;
-    CHECK(*i == '1');
-
-    auto j = x.safeEnd();
-    CHECK(j != i);
-    CHECK(j == x.safeEnd());
-
-    x.append("abc"_b);
-    CHECK(j != x.safeEnd());
-    CHECK(*j == 'a');
-
-    ++j;
-    CHECK(j != x.safeEnd());
-    ++j;
-    CHECK(j != x.safeEnd());
-    ++j;
-    CHECK(j == x.safeEnd());
-
-    j += 5;
-    CHECK_THROWS_AS(*j, InvalidIterator);
-    x.append("1234567890"_b);
-    CHECK(*j == '6');
-
-    x = Stream(""_b);
-    i = x.safeBegin();
-    CHECK_THROWS_AS(*i, InvalidIterator);
-    x.append("1"_b);
-    CHECK(*i == '1');
-
-    CHECK_THROWS_AS((void)(*j == '6'), InvalidIterator); // j now invalid.
+TEST_CASE("iteration") {
+    SUBCASE("sees data") {
+        auto x = Stream("12345"_b);
+
+        std::string s;
+        for ( auto i : x )
+            s += i;
+
+        CHECK_EQ(s, "12345");
+    }
+
+    SUBCASE("see data updates") {
+        auto x = Stream("12345"_b);
+        x.append("1234567890"_b);
+        x.append("1234567890"_b);
+        x.append("1234567890"_b);
+        x.append("1234567890"_b);
+
+        std::string s;
+        for ( auto i : x )
+            s += i;
+
+        CHECK_EQ(s, "123451234567890123456789012345678901234567890");
+    }
+
+    SUBCASE("equality") {
+        auto x = Stream("123451234567890123456789012345678901234567890"_b);
+        auto i = x.safeBegin();
+        i += 7;
+        CHECK_EQ(*i, '3');
+        i += 7;
+        CHECK_EQ(*i, '0');
+        i += 1;
+        CHECK_EQ(*i, '1');
+
+        auto j = x.safeEnd();
+        CHECK_NE(j, i);
+        CHECK_EQ(j, x.safeEnd());
+
+        x.append("abc"_b);
+        CHECK_NE(j, x.safeEnd());
+        CHECK_EQ(*j, 'a');
+
+        ++j;
+        CHECK_NE(j, x.safeEnd());
+        ++j;
+        CHECK_NE(j, x.safeEnd());
+        ++j;
+        CHECK_EQ(j, x.safeEnd());
+    }
+
+    SUBCASE("rangecheck") {
+        auto x = Stream("123"_b);
+        auto i = x.begin();
+
+        i += 5; // Points beyond the end of the available data.
+        CHECK_THROWS_AS(*i, InvalidIterator);
+
+        x.append("4567890"_b);
+        CHECK_EQ(*i, '6'); // Enough data available now.
+    }
+
+    SUBCASE("lifetime bound by underlying stream") {
+        auto j = Stream().begin();
+        CHECK_THROWS_AS((void)(*j == '6'), InvalidIterator); // j now invalid.
+    }
+
+    SUBCASE("invariant when data added") {
+        auto s = Stream("0123"_b);
+        auto i0 = s.begin();
+        auto i1 = i0 + 1;
+        REQUIRE_EQ(*i0, '0');
+        REQUIRE_EQ(*i1, '1');
+
+        s.append("456789"_b);
+
+        CHECK_EQ(*i0, '0');
+        CHECK_EQ(*i1, '1');
+    }
+
+    SUBCASE("difference") {
+        const auto [s, before_begin] = []() {
+            auto s = Stream(" 123"_b);
+            const auto before_begin = s.begin();
+
+            s.trim(before_begin + 1);
+            REQUIRE_EQ(s, "123"_b);
+
+            return std::make_tuple(s, before_begin);
```

Comment:
Isn''t `before_begin` just becoming an expired iterator here because `s` won't survive the return?

Reply:
Good catch! I added a `REQUIRE_FALSE(before_begin.isExpired())` which failed with the code in this form, but succeeds if we `std::move(s)` into the returned tuple.

Reply:
ok, nevermind then